scripts/internal/prune_size_model.py

```diff
@@ -0,0 +1,310 @@
+
+        next_larger_num_xgrams = cur_target_num_xgrams
+        while left <= right - tolerance:
+            next_threshold = (left + right) / 2
+            modeled_next_num_xgrams = self.GetModeledNextNumXgrams(next_threshold)
+
+            if modeled_next_num_xgrams < cur_target_num_xgrams:
+                right = next_threshold
+            elif modeled_next_num_xgrams > cur_target_num_xgrams:
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                left = next_threshold
+            else: # modeled_next_num_xgrams == cur_target_num_xgrams, this will probably not happan
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                break
+        if left > right - tolerance:
+            # the while loop is not breaked by the else clause,
+            # so we make sure the modeled_next_num_xgrams >= cur_target_num_xgrams
+            next_threshold = left
+
+        if math.fabs(cur_target_num_xgrams - next_larger_num_xgrams) > 0.2 * cur_target_num_xgrams:
+            LogMessage("Warning: final modeled_next_num_xgrams[{0}] is far "
+                       "from intermediate_target_num_xgrams{1}".format(
+                           next_larger_num_xgrams, cur_target_num_xgrams))
+
+        return (next_threshold, next_larger_num_xgrams)
+
+    def GetModeledNextNumXgrams(self, next_threshold):
+        """
+        This function is a model of how we think the next num-xgrams
+        will vary with the chosen next threshold.
+        We predict the next_num_xgrams based on num_xgrams changed of
+        previous iteration and changes of threshold.
+        """
+
+        prev_num_xgrams = self.GetPrevNumXgrams()
+        cur_threshold = self.GetCurThreshold()
+        cur_num_xgrams = self.GetCurNumXgrams()
+
+        # First predict the num-xgrams we think we'll get if we prune again
+        # with the same threshold 'cur_threshold'. The basic heuristic is
+        # that the num-xgrams will decrease by a factor no greater than
+        # 1.5, and no greater than the cube root of the factor by which
+        # the num-xgrams decreased on the previous iteration.
+        assert prev_num_xgrams >= cur_num_xgrams
+        prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) ** (1.0/3.0)
+        baseline_decrease_factor = 1.0 / 1.5
+        # choose the factor from these two, that's closest to one.
+        predicted_decrease_factor = max(prev_change_factor, baseline_decrease_factor)
+
+        # the following gives us the predicted num-xgrams if we were
+        # to prune again with the same threshold 'cur_threshold'.
+        predicted_num_xgrams_if_repeat = predicted_decrease_factor * cur_num_xgrams
+        assert next_threshold >= cur_threshold
+
+        predicted_extra_factor = (next_threshold / cur_threshold) ** (self.xgrams_change_power)
+        return predicted_num_xgrams_if_repeat * predicted_extra_factor
+
+    def AdjustModelForOvershoot(self):
```

Comment:
> Firstly, you are doing some things in this function that don't have anything to do with adjusting the model.
> You should have a function Backtrack() that will do the actual backtracking. Adjusting the model is a separate issue.

I don't quite understand this, this function modifies the value of xgrams_change_power, which is a parameter of the model. Isn't it doing adjusting model?
The actual backtracking is done in the GetNextAction function by simply returning a appropriate backtrack_iter to the caller.

> For backtracking, if the two most recent thresholds were identical, then you should remove the two most recent elements from the history [i.e. backtrack twice]. If the history would then be empty, then the user-specified initial threshold was too high and you can exit with failure.

It would remove two most recent iterations if their threshold are the same. Actually, I change this to remove more iterations if their threshold is all the same, since we could sometimes run for 3 or more iterations with a same threshold until we find it need to backtrack.

The newer codes are not pushed to this branch, because i'm not sure whether it is the right way to do backtrack. I will push them later after I tried your new modifications.

There are two other changes i made. First, I store the parameters of model(i.e. xgrams_change_power for now) for every iteration instead of as a global variable. Since when backtrack to a iteration, we may have done many backtracks from a different starting iteration, and this will make the model too less aggressive if we continue to use the adjusted parameter.

Second, I think the model could be more aggressive instead of less, when it is backtracking from the next two or more iterations. Because we already know that if we use the same threshold, it will result to a model whose size is larger than target. By using a larger threshold, we could be closer to target and may reach the target in fewer iterations.
 
PS: Your comment seems to be truncated, i could only see this at the end:
> Next, the way you are adjusting the model is not very good.
> You need to adjust the